lib/base/json.hpp

```diff
@@ -4,14 +4,270 @@
+ * The supported stream types include std::ostream and AsioStreamAdapter. The nlohmann/json library already
+ * provides a full support for the former stream type, while the latter is fully implemented by our own and
+ * satisfies the @c nlohmann::detail::output_adapter_protocol<> interface. It is used for writing the produced
+ * JSON directly to an Asio either TCP or TLS stream and doesn't require any additional buffering other than
+ * the one used by the Asio buffered_stream<> class internally.
+ *
+ * The JSON encoder generates most of the low level JSON tokens, but it still relies on the already existing
+ * @c nlohmann::detail::serializer<> class to dump numbers and ascii validated JSON strings. This means that the
+ * encoder doesn't perform any kind of JSON validation or escaping on its own, but simply delegates all this kind
+ * of work to serializer<>. However, Strings are UTF-8 validated beforehand using the @c Utility::ValidateUTF8()
+ * function and only the validated (copy of the original) String is passed to the serializer.
+ *
+ * The generated JSON can be either prettified or compact, depending on your needs. The prettified JSON object
+ * is indented with 4 spaces and grows linearly with the depth of the object tree.
+ *
+ * @ingroup base
+ */
+class JsonEncoder
+{
+public:
+	explicit JsonEncoder(std::string& output, bool prettify = false);
+	explicit JsonEncoder(std::basic_ostream<char>& stream, bool prettify = false);
+	explicit JsonEncoder(nlohmann::detail::output_adapter_t<char> stream, bool prettify = false);
+
+	/**
+	 * Encodes a single value into JSON and writes it to the underlying output stream.
+	 *
+	 * This function is a convenience wrapper around the @c EncodeImpl() method, which allows encoding
+	 * a single value without having to specify the type explicitly. It can be used to encode any type
+	 * that is supported by the @c JsonEncoder, such as Value or Ctrls.
+	 *
+	 * It can be used to incrementally build a JSON object or array by adding elements one by one.
+	 *
+	 * @tparam T The type of the value to encode.
+	 *
+	 * @param value The value to encode into JSON format.
+	 *
+	 * @return A reference to the current JsonEncoder instance for method chaining.
+	 */
+	template<typename T>
+	void Encode(T&& value)
+	{
+		// If we are currently inside an array context, never use the generic EncodeImpl() method, but rather
+		// call EncodeItem() to add the item as an element to the array. Otherwise, this will mess up the JSON,
+		// since the generic EncodeImpl() isn't aware of the current context.
+		if (!m_CtxStack.empty() && m_CtxStack.top().type == CtxType::Array) {
+			EncodeItem(std::forward<T>(value));
+			return;
+		}
+		EncodeImpl(std::forward<T>(value));
+		// If we are at the top level of the JSON object and prettifying is enabled, we need
+		// to end the JSON with a newline character to ensure that the output is properly formatted.
+		if (m_Indent == 0 && m_Pretty) {
+			EncodeImpl(NewLine);
+		}
```

Comment:
```suggestion
		EncodeImpl(std::forward<T>(value));
		// If we are at the top level of the JSON object and prettifying is enabled, we need
		// to end the JSON with a newline character to ensure that the output is properly formatted.
		if (m_Pretty) {
			EncodeImpl(NewLine);
		}
```

And at the top you could just assert `m_CtxStack.empty()`.